src/platform/PICO/bus_i2c_pico.c

```diff
@@ -195,14 +258,181 @@ bool i2cBusy(I2CDevice device, bool *error)
         *error = 0;
     }
 
-    // TODO check: If we are using DMA (for a sequence of transfers?), then we will need to
-    // protect against that being in progress
+    // Check if we have a transfer in progress via our state machine
+    i2c_context_t *context = &i2c_contexts[device];
+    if (context->state != I2C_STATE_IDLE) {
+        return true;
+    }
 
-    // Read the IC_STATUS register
+    // Also check hardware status
     uint32_t status_reg = port->hw->status;
+    return (status_reg & I2C_IC_STATUS_ACTIVITY_VALUE_ACTIVE) != 0;
+}
+
+static void i2c_handle_error(i2c_context_t *i2c_context, uint32_t intr_stat)
+{
+    i2c_hw_t *hw = i2c_context->i2c->hw;
 
-    // The bit for (combined master/slave) ACTIVITY is (1 << 0).
-    return (status_reg & (1 << 0)) != 0;
+    // Enable only default interrupts
+    hw->intr_mask = DEF_I2C_INTR;
+    i2c_context->state = I2C_STATE_IDLE;
+    i2c_context->intr_error_stat = intr_stat & (DEF_I2C_INTR);
+    i2cErrorCount++;
+}
+// Common I2C IRQ handler
+static void i2c_irq_handler(i2c_context_t *i2c_context)
```

Comment:
With 16 level deep FIFO, is seems that most transactions can be simply written into fifo in fire-and-forget mode. In interrupt, only long read/write operations need to be handled - refill fifo with WR commands/store received data.

Reply:
Thanks for the prompt to look into this.